### ISSUE TITLE
Using X as a parameter to summon X cards, deal X damage, etc.

### ISSUE THREAD

**Issue by Zomis** [Cardshifter/Cardshifter#240]
For example:

summon X Bodyman creatures, where X is half the total number of creatures you control (rounded down)

So if you have 4 creatures, you summon 2 new ones. If you have 7 creatures, you summon 3.

Another example can be:

Deal X damage to random opponent creature, where X is the highest attack value of creatures you control

**Comment by jay1148** [comment]
I like this. I have a location card whose attack and health are equal to the (X) number of a certain class of cards currently controlled. I believe that is along the same lines. I think that should be implemented if possible.

**Comment by jacwah** [comment]
The effect parameters are not required to be literals. We don't need to implement a new DSL syntax to allow this sort of behavior. Of course, it *could* be easier to understand for non-programmers to understand with a simple syntax, if there are any good suggestions.

All that is needed to meet the requirements of variable arguments would be to define an intuitive API for users to query.

```
summon player.creatures.count of 'Bodyman' to 'you' zone 'Hand'
```

In the example, `player` is an object containing data about the current player, and `creatures` is a list of all creatures they control.

For more complex behavior, I think a bit of programming will be necessary as we can't define a DSL keyword for all possible things people might come up with.

```
damage player.creatures.collect({it.attack}).max() on targets
```

This can also be done by defining X separately.

```
def X = player.creatures.collect({it.attack}).max()
damage X on targets
```

If we define an empty `X` variable, the `def` can be dropped.

**Comment by Zomis** [comment]
I agree that a bit of programming is necessary, however, at the moment all the lines such as

`summon 3 of 'Bodyman' to 'you' zone 'Hand'`

are evaluated once, during initialization of the card model. It is then transformed into two things: A Description (human readable) and an Effect (code). I agree that allowing some programming is a good thing here (which is possible by using some kind of closure / callback function that can take parameters and return a result), but we also need a way to provide a description (define manual description or try to generate the description automatically?)

**Comment by Phrancis** [comment]
I think a manual description would be OK/good, something that can override the default generated description, if/when desired. That should probably be its own issue though.

**Comment by jacwah** [comment]
Regarding descriptions, the more flexible the code is, l the harder it will be to generate descriptions. Using using arbitrary stand library methods, like in my example, would make it impossible. We could alternatively provide a `Collection` subclass with certain wrapped methods like `Colletion.collect` and `Collection.count`, though I can see this getting out of hand.

One option, like already mentioned, is to simply not generate a description at all for these cards. Some changes would have to be done to avoid evaluating `X` when the description is generated. A separate property describing the card would then be provided by the modder.

A middle-of-the-road approach could perhaps be to let the modder provide a human readble desciption only of `X`, and include this in the auto-generated text.

```
X.describe "the highest attack value of the creates you control"
X.effect { player.creatures.collect({it.attack}).max() }
damage X on targets
```

In this example, `X` would be an object that can be resolved both as a closure and a string, resulting in the description "Deal the highest attack value of the creates you control damage to random opponent creature". The generation can of course be tweaked to perhaps read "Deal damage equal to the highest attack value of the creates you control to random opponent creature".
